File: src/logic/riskEngine.js
/**
 * TrialGuard AI — Deterministic Risk Engine
 * 
 * Scores individual patients, clinical sites, and overall trial safety & compliance.
 * Implements severity weights (Critical: 15, Major: 10, Minor: 4, Admin: 1)
 * and risk modifiers (+5 repeated deviation, +5 multiple categories).
 * Normalizes site risk to a 0–100 scale.
 */

import { PROTOCOL_CONFIG } from "../data/protocolConfig.js";

export function calculatePatientRisk(patient, deviations, protocol = PROTOCOL_CONFIG) {
  const patientDevs = deviations.filter((d) => d.patientId === patient.id);
  
  if (patientDevs.length === 0) {
    return {
      patientId: patient.id,
      deviationCount: 0,
      basePoints: 0,
      repeatedModifier: 0,
      multipleCategoryModifier: 0,
      totalRiskScore: 0,
      riskBand: "Low",
      compliancePercentage: 100,
      status: patient.status || "Active",
      deviations: []
    };
  }

  // Base points from severity weights
  const basePoints = patientDevs.reduce((sum, d) => sum + (d.severityWeight || 0), 0);

  // Modifier: repeated deviations (+5 if patient has > 1 deviation)
  const repeatedModifier = patientDevs.length > 1 ? protocol.riskModifiers.repeatedDeviation : 0;

  // Modifier: multiple deviation categories (+5 if patient has deviations in > 1 category)
  const distinctCategories = new Set(patientDevs.map((d) => d.category));
  const multipleCategoryModifier = distinctCategories.size > 1 ? protocol.riskModifiers.multipleCategories : 0;

  const totalRiskScore = basePoints + repeatedModifier + multipleCategoryModifier;

  // Derive compliance %
  const compliancePercentage = Math.max(0, Math.min(100, Math.round(100 - totalRiskScore * 1.8)));

  let riskBand = "Low";
  if (totalRiskScore > 35) {
    riskBand = "High";
  } else if (totalRiskScore > 12) {
    riskBand = "Medium";
  }

  return {
    patientId: patient.id,
    deviationCount: patientDevs.length,
    basePoints,
    repeatedModifier,
    multipleCategoryModifier,
    totalRiskScore,
    riskBand,
    compliancePercentage,
    status: patient.status || "Active",
    deviations: patientDevs
  };
}

export function calculateSiteRisk(site, patients, deviations, protocol = PROTOCOL_CONFIG) {
  const sitePatients = patients.filter((p) => p.siteId === site.id);
  const siteDeviations = deviations.filter((d) => d.siteId === site.id);
  const patientCount = sitePatients.length || 1;

  // Calculate each patient's risk profile
  const patientProfiles = sitePatients.map((p) => calculatePatientRisk(p, deviations, protocol));
  const affectedPatients = patientProfiles.filter((p) => p.deviationCount > 0);

  // Sum of total risk points across all patients at this site
  const rawRiskPoints = patientProfiles.reduce((sum, p) => sum + p.totalRiskScore, 0);

  // Counts by severity
  const criticalDevs = siteDeviations.filter((d) => d.severity === "Critical");
  const majorDevs = siteDeviations.filter((d) => d.severity === "Major");
  const minorDevs = siteDeviations.filter((d) => d.severity === "Minor");
  const adminDevs = siteDeviations.filter((d) => d.severity === "Administrative");

  // Normalized 0–100 site risk formula based on density, proportion of affected cohort, and critical density
  const density = rawRiskPoints / patientCount;
  const affectedRatio = affectedPatients.length / patientCount;
  const criticalWeight = criticalDevs.length * 2.6;
  const majorWeight = majorDevs.length * 1.8;

  let normalizedScore = Math.round(density * 3.7 + affectedRatio * 18 + criticalWeight + majorWeight);
  // Clamp between 0 and 100
  normalizedScore = Math.max(0, Math.min(100, normalizedScore));

  // Determine band according to hackathon criteria (0-30 Low, 31-60 Medium, 61-100 High)
  let riskBand = "Low";
  if (normalizedScore >= 61) {
    riskBand = "High";
  } else if (normalizedScore >= 31) {
    riskBand = "Medium";
  }

  // Key risk drivers analysis
  const riskDrivers = [];
  if (criticalDevs.length > 0) {
    const types = [...new Set(criticalDevs.map(d => d.category))].join(", ");
    riskDrivers.push(`${criticalDevs.length} Critical Deviation${criticalDevs.length > 1 ? "s" : ""} (${types})`);
  }
  if (majorDevs.length >= 3) {
    riskDrivers.push(`High Major deviation density (${majorDevs.length} major violations recorded)`);
  }
  const multiViolationPts = patientProfiles.filter((p) => p.deviationCount > 1).length;
  if (multiViolationPts > 0) {
    riskDrivers.push(`${multiViolationPts} patient${multiViolationPts > 1 ? "s" : ""} with recurrent multi-category violations`);
  }
  if (affectedRatio > 0.3) {
    riskDrivers.push(`Elevated cohort contamination: ${(affectedRatio * 100).toFixed(0)}% of site subjects affected`);
  }
  if (riskDrivers.length === 0) {
    riskDrivers.push("Good protocol adherence; minimal isolated discrepancies");
  }

  // Deviation category breakdown
  const categoryBreakdown = siteDeviations.reduce((acc, d) => {
    acc[d.category] = (acc[d.category] || 0) + 1;
    return acc;
  }, {});

  return {
    siteId: site.id,
    siteCode: site.code,
    siteName: site.name,
    location: site.location,
    pi: site.pi,
    cra: site.cra,
    patientCount,
    affectedPatientCount: affectedPatients.length,
    deviationCount: siteDeviations.length,
    rawRiskPoints,
    score: normalizedScore,
    riskBand,
    criticalCount: criticalDevs.length,
    majorCount: majorDevs.length,
    minorCount: minorDevs.length,
    adminCount: adminDevs.length,
    riskDrivers,
    categoryBreakdown,
    deviations: siteDeviations,
    patientProfiles
  };
}

export function calculateTrialMetrics(sites, patients, deviations, protocol = PROTOCOL_CONFIG) {
  const siteRiskList = sites.map((s) => calculateSiteRisk(s, patients, deviations, protocol));
  
  // Rank sites by risk score descending
  siteRiskList.sort((a, b) => b.score - a.score);

  const totalPatients = patients.length || 1;
  const totalDeviations = deviations.length;
  const criticalDeviations = deviations.filter((d) => d.severity === "Critical").length;
  const majorDeviations = deviations.filter((d) => d.severity === "Major").length;
  const openDeviations = deviations.filter((d) => d.status !== "Resolved").length;

  const highRiskSites = siteRiskList.filter((s) => s.riskBand === "High");
  const mediumRiskSites = siteRiskList.filter((s) => s.riskBand === "Medium");

  // Overall trial risk score (weighted average of site risks)
  const trialRiskScore = Math.round(
    siteRiskList.reduce((sum, s) => sum + s.score, 0) / (siteRiskList.length || 1)
  );

  let trialRiskBand = "Low";
  if (trialRiskScore >= 61) trialRiskBand = "High";
  else if (trialRiskScore >= 31) trialRiskBand = "Medium";

  // Overall compliance rate
  const totalEvaluations = totalPatients * 5; // ~5 key checkpoints per patient
  const compliancePercentage = Math.max(0, Number(((1 - (totalDeviations / totalEvaluations)) * 100).toFixed(1)));

  return {
    trialId: protocol.trialId,
    trialName: protocol.trialName,
    phase: protocol.phase,
    totalPatients,
    totalSites: sites.length,
    totalDeviations,
    openDeviations,
    criticalDeviations,
    majorDeviations,
    trialRiskScore,
    trialRiskBand,
    compliancePercentage,
    highRiskSitesCount: highRiskSites.length,
    mediumRiskSitesCount: mediumRiskSites.length,
    siteRiskList
  };
}

File: src/backend/app/services/risk_engine.py
"""TrialGuard AI — Deterministic Site Risk Engine & Trend Predictor.

Calculates objective site-level and patient-level risk scores directly from protocol deviations:
- Severity weights: Critical=15, Major=10, Minor=4, Administrative=1
- Modifiers: +5 for repeated deviations at site, +5 for multiple deviation categories
- Normalized 0–100 scale: Low (0–30), Medium (31–60), High (61–100)
- Trend determination: Improving, Stable, Worsening based on deviation velocity
- Heuristic 30-day predicted risk with transparent clinical rationale
- Dynamic top risk driver extraction directly from actual deviation records
"""
from typing import List, Dict, Any, Optional
from ..models.patient import Patient
from ..models.deviation import Deviation
from ..models.trial import (
    ProtocolConfig,
    PatientProfile,
    SiteRiskSummary,
    TrialMetrics,
    SiteTrendPrediction,
    SiteTrendHistoryPoint,
    TopRiskDriverItem,
)
from ..data.protocol import PROTOCOL_CONFIG

# Deterministic normalization reference scale (points corresponding to extreme site risk)
MAX_SITE_RISK_SCALE = 225.0

def calculate_patient_risk(
    patient: Patient,
    deviations: List[Deviation],
    protocol: ProtocolConfig = PROTOCOL_CONFIG
) -> PatientProfile:
    patient_devs = [d for d in deviations if d.patientId == patient.id]

    if not patient_devs:
        return PatientProfile(
            patientId=patient.id,
            deviationCount=0,
            basePoints=0,
            repeatedModifier=0,
            multipleCategoryModifier=0,
            totalRiskScore=0,
            riskBand="Low",
            compliancePercentage=100.0,
            status=patient.status,
            deviations=[],
        )

    base_points = sum(d.severityWeight for d in patient_devs)
    repeated_modifier = protocol.riskModifiers.get("repeatedDeviation", 5) if len(patient_devs) > 1 else 0
    distinct_categories = {d.category for d in patient_devs}
    multiple_category_modifier = protocol.riskModifiers.get("multipleCategories", 5) if len(distinct_categories) > 1 else 0

    total_risk_score = base_points + repeated_modifier + multiple_category_modifier
    compliance_percentage = max(0.0, min(100.0, float(round(100 - total_risk_score * 1.8, 1))))

    if total_risk_score > 35:
        risk_band = "High"
    elif total_risk_score > 12:
        risk_band = "Medium"
    else:
        risk_band = "Low"

    return PatientProfile(
        patientId=patient.id,
        deviationCount=len(patient_devs),
        basePoints=base_points,
        repeatedModifier=repeated_modifier,
        multipleCategoryModifier=multiple_category_modifier,
        totalRiskScore=total_risk_score,
        riskBand=risk_band,
        compliancePercentage=compliance_percentage,
        status=patient.status,
        deviations=patient_devs,
    )

def compute_site_trend_and_prediction(
    score: int,
    risk_band: str,
    critical_count: int,
    major_count: int,
    repeated_count: int,
    affected_count: int,
    patient_count: int,
) -> SiteTrendPrediction:
    """Deterministically computes trend (Improving, Stable, Worsening) and 30-day predicted risk."""
    pt_count = patient_count if patient_count > 0 else 1
    infection_ratio = affected_count / pt_count

    # Trend logic based on deviation velocity and recurrent high-severity clusters
    if critical_count >= 2 or (critical_count >= 1 and major_count >= 3) or repeated_count >= 5 or infection_ratio > 0.4:
        trend = "Worsening"
        projected_change = int(round(5 + critical_count * 1.6 + major_count * 0.6))
        prediction_direction = "Increasing"
        rationale = (
            f"Active cluster of {critical_count} critical and {major_count} major deviations "
            f"with {repeated_count} repeated occurrences indicates systemic site compliance lapses. "
            "Risk is projected to elevate unless corrective action is enacted immediately."
        )
    elif score <= 15 and critical_count == 0 and major_count <= 1:
        trend = "Improving"
        projected_change = -min(score, 3)
        prediction_direction = "Decreasing"
        rationale = (
            "Consistent protocol adherence with zero critical safety violations. "
            "Site is maintaining high-fidelity clinical procedures."
        )
    else:
        trend = "Stable"
        projected_change = 0
        prediction_direction = "Stable"
        rationale = (
            "Isolated administrative or moderate deviations with controlled event velocity; "
            "risk trajectory remains steady under standard monitoring oversight."
        )

    predicted_score = max(0, min(100, score + projected_change))

    if predicted_score >= 61:
        predicted_band = "High"
    elif predicted_score >= 31:
        predicted_band = "Medium"
    else:
        predicted_band = "Low"

    explanation = (
        f"Current risk score {score} ({risk_band}) projected to {prediction_direction.lower()} "
        f"to {predicted_score} ({predicted_band}) over 30 days. {rationale}"
    )

    history = [
        SiteTrendHistoryPoint(week="Wk -3", score=max(0, score - int(round(projected_change * 1.8)))),
        SiteTrendHistoryPoint(week="Wk -2", score=max(0, score - int(round(projected_change * 1.2)))),
        SiteTrendHistoryPoint(week="Wk -1", score=max(0, score - int(round(projected_change * 0.5)))),
        SiteTrendHistoryPoint(week="Current", score=score),
        SiteTrendHistoryPoint(week="Wk +2 (Est)", score=int(round(score + projected_change * 0.5)), isForecast=True),
        SiteTrendHistoryPoint(week="Wk +4 (Est)", score=predicted_score, isForecast=True),
    ]

    return SiteTrendPrediction(
        trend=trend,
        predictedScore=predicted_score,
        predictedBand=predicted_band,
        projectedChange=projected_change,
        predictionRationale=rationale,
        currentRisk=score,
        predictedRisk=predicted_score,
        predictionDirection=prediction_direction,
        explanation=explanation,
        history=history,
        disclaimer="Deterministic heuristic trajectory based on deviation velocity & cluster recurrence.",
    )

compute_site_trend = compute_site_trend_and_prediction

def extract_site_risk_drivers(
    site_deviations: List[Deviation],
    affected_count: int,
    patient_count: int,
) -> tuple[List[str], List[TopRiskDriverItem]]:
    """Identifies the biggest contributors to site risk derived directly from actual deviations."""
    drivers: List[str] = []
    top_items: List[TopRiskDriverItem] = []

    # 1. Prohibited Concomitant Medication
    prohibited_devs = [d for d in site_deviations if d.type == "PROHIBITED_CONMED"]
    if prohibited_devs:
        med_names = sorted(list({
            d.actual.split()[2] if len(d.actual.split()) > 2 else "CYP3A4 inhibitor"
            for d in prohibited_devs
        }))
        d_text = f"Prohibited medication administration ({len(prohibited_devs)} incident{'s' if len(prohibited_devs) > 1 else ''}: {', '.join(med_names)})"
        drivers.append(d_text)
        top_items.append(TopRiskDriverItem(driver=d_text, severity="Critical", count=len(prohibited_devs)))

    # 2. Investigational Product Dosing Deviations
    dose_devs = [d for d in site_deviations if d.type == "INCORRECT_DOSE"]
    if dose_devs:
        has_crit = any(d.severity == "Critical" for d in dose_devs)
        d_text = f"Investigational product dose deviations ({len(dose_devs)} subject{'s' if len(dose_devs) > 1 else ''} received non-protocol dose)"
        drivers.append(d_text)
        top_items.append(TopRiskDriverItem(driver=d_text, severity="Critical" if has_crit else "Major", count=len(dose_devs)))

    # 3. Repeated Late / Missed Visits
    window_devs = [d for d in site_deviations if d.type in ["VISIT_WINDOW_EXCEEDED", "MISSED_VISIT", "VISIT_TOO_EARLY"]]
    if len(window_devs) >= 2:
        d_text = f"Repeated late / missed visits ({len(window_devs)} visit window lapses across schedule)"
        drivers.append(d_text)
        top_items.append(TopRiskDriverItem(driver=d_text, severity="Major", count=len(window_devs)))
    elif len(window_devs) == 1:
        d_text = f"Isolated visit timing discrepancy ({window_devs[0].type.replace('_', ' ').title()})"
        drivers.append(d_text)
        top_items.append(TopRiskDriverItem(driver=d_text, severity=window_devs[0].severity, count=1))

    # 4. Missing Mandatory Laboratory Assessments
    lab_devs = [d for d in site_deviations if d.type == "MISSING_MANDATORY_LAB"]
    if lab_devs:
        d_text = f"Missing mandatory laboratory assessments ({len(lab_devs)} CBC panel{'s' if len(lab_devs) > 1 else ''} omitted before Visit 3)"
        drivers.append(d_text)
        top_items.append(TopRiskDriverItem(driver=d_text, severity="Major", count=len(lab_devs)))

    # 5. Omitted Procedures / Incomplete Vitals
    proc_devs = [d for d in site_deviations if d.type in ["PROCEDURE_OMITTED", "INCOMPLETE_VITALS"]]
    if proc_devs:
        has_major = any(d.severity == "Major" for d in proc_devs)
        d_text = f"Protocol procedure / vital sign omissions ({len(proc_devs)} checklist omission{'s' if len(proc_devs) > 1 else ''})"
        drivers.append(d_text)
        top_items.append(TopRiskDriverItem(driver=d_text, severity="Major" if has_major else "Administrative", count=len(proc_devs)))

    # 6. Elevated Cohort Impact
    if patient_count > 0 and (affected_count / patient_count) > 0.35:
        pct = int(round((affected_count / patient_count) * 100))
        d_text = f"Elevated cohort impact ({pct}% of site subjects affected by deviations)"
        drivers.append(d_text)
        top_items.append(TopRiskDriverItem(driver=d_text, severity="Major", count=affected_count))

    if not drivers:
        d_text = "Good protocol adherence; zero compliance deviations detected"
        drivers.append(d_text)
        top_items.append(TopRiskDriverItem(driver=d_text, severity="Administrative", count=0))

    return drivers, top_items

def calculate_site_risk(
    site: Any,
    patients: List[Patient],
    deviations: List[Deviation],
    protocol: ProtocolConfig = PROTOCOL_CONFIG,
    include_details: bool = True
) -> SiteRiskSummary:
    """Calculates deterministic site risk score, metrics, trend, and drivers from actual deviations."""
    site_patients = [p for p in patients if p.siteId == site.id]
    site_deviations = [d for d in deviations if d.siteId == site.id]
    patient_count = len(site_patients) or 1

    patient_profiles = [calculate_patient_risk(p, deviations, protocol) for p in site_patients]
    affected_patients = [p for p in patient_profiles if p.deviationCount > 0]
    affected_count = len(affected_patients)

    # 1. Severity Counts
    critical_devs = [d for d in site_deviations if d.severity == "Critical"]
    major_devs = [d for d in site_deviations if d.severity == "Major"]
    minor_devs = [d for d in site_deviations if d.severity == "Minor"]
    admin_devs = [d for d in site_deviations if d.severity == "Administrative"]

    critical_count = len(critical_devs)
    major_count = len(major_devs)
    minor_count = len(minor_devs)
    admin_count = len(admin_devs)
    deviation_count = len(site_deviations)

    # 2. Severity Weighting
    severity_sum = (
        critical_count * protocol.severityWeights.get("Critical", 15)
        + major_count * protocol.severityWeights.get("Major", 10)
        + minor_count * protocol.severityWeights.get("Minor", 4)
        + admin_count * protocol.severityWeights.get("Administrative", 1)
    )

    # 3. Repeated Deviations Calculation
    type_counts: Dict[str, int] = {}
    for d in site_deviations:
        type_counts[d.type] = type_counts.get(d.type, 0) + 1
    repeated_deviation_count = sum(max(0, count - 1) for count in type_counts.values())

    repeated_bonus = protocol.riskModifiers.get("repeatedDeviation", 5) if repeated_deviation_count > 0 else 0

    # 4. Multiple Categories Calculation
    category_breakdown: Dict[str, int] = {}
    for d in site_deviations:
        category_breakdown[d.category] = category_breakdown.get(d.category, 0) + 1
    number_of_categories = len(category_breakdown)

    multi_category_bonus = protocol.riskModifiers.get("multipleCategories", 5) if number_of_categories > 1 else 0

    # 5. Raw Risk Points & 0–100 Normalization
    raw_risk_points = severity_sum + repeated_bonus + multi_category_bonus

    if raw_risk_points == 0:
        normalized_score = 0
    else:
        # Scale to 0-100 benchmark
        normalized_score = min(100, max(1, int(round((raw_risk_points / MAX_SITE_RISK_SCALE) * 100))))

    # 6. Risk Level / Band
    if normalized_score >= 61:
        risk_band = "High"
    elif normalized_score >= 31:
        risk_band = "Medium"
    else:
        risk_band = "Low"

    # 7. Recent Deviations (within last 14 days or study period >= 2026-02-08)
    recent_dev_count = len([
        d for d in site_deviations
        if d.detectedAt >= "2026-02-08" or (d.date and d.date >= "2026-02-08")
    ])

    # 8. Top Risk Drivers
    risk_drivers, top_driver_items = extract_site_risk_drivers(site_deviations, affected_count, patient_count)

    # 9. Dynamic Trend & Prediction
    trend_prediction = compute_site_trend_and_prediction(
        score=normalized_score,
        risk_band=risk_band,
        critical_count=critical_count,
        major_count=major_count,
        repeated_count=repeated_deviation_count,
        affected_count=affected_count,
        patient_count=patient_count,
    )

    return SiteRiskSummary(
        siteId=site.id,
        siteCode=site.code,
        siteName=site.name,
        location=site.location,
        pi=site.pi,
        cra=site.cra,
        patientCount=patient_count,
        affectedPatientCount=affected_count,
        deviationCount=deviation_count,
        rawRiskPoints=raw_risk_points,
        score=normalized_score,
        riskBand=risk_band,
        criticalCount=critical_count,
        majorCount=major_count,
        minorCount=minor_count,
        adminCount=admin_count,
        repeatedDeviationCount=repeated_deviation_count,
        numberOfCategories=number_of_categories,
        categoryCount=number_of_categories,
        recentDeviationCount=recent_dev_count,
        riskDrivers=risk_drivers,
        topDrivers=top_driver_items,
        topRiskDrivers=top_driver_items,
        categoryBreakdown=category_breakdown,
        deviations=site_deviations if include_details else None,
        patientProfiles=patient_profiles if include_details else None,
        trend=trend_prediction,
    )

def calculate_trial_metrics(
    sites: List[Any],
    patients: List[Patient],
    deviations: List[Deviation],
    protocol: ProtocolConfig = PROTOCOL_CONFIG
) -> TrialMetrics:
    site_risk_list = [
        calculate_site_risk(s, patients, deviations, protocol, include_details=True)
        for s in sites
    ]
    site_risk_list.sort(key=lambda s: s.score, reverse=True)

    total_patients = len(patients) or 1
    total_deviations = len(deviations)
    critical_devs = len([d for d in deviations if d.severity == "Critical"])
    major_devs = len([d for d in deviations if d.severity == "Major"])
    open_devs = len([d for d in deviations if d.status != "Resolved"])

    high_risk_sites = [s for s in site_risk_list if s.riskBand == "High"]
    medium_risk_sites = [s for s in site_risk_list if s.riskBand == "Medium"]

    trial_risk_score = int(round(sum(s.score for s in site_risk_list) / (len(site_risk_list) or 1)))

    if trial_risk_score >= 61:
        trial_risk_band = "High"
    elif trial_risk_score >= 31:
        trial_risk_band = "Medium"
    else:
        trial_risk_band = "Low"

    total_evaluations = total_patients * 5
    compliance_pct = max(0.0, float(round((1.0 - (total_deviations / float(total_evaluations))) * 100.0, 1)))

    from .explanation_service import generate_trial_executive_insight
    highest_site = site_risk_list[0] if site_risk_list else None
    executive_insight = generate_trial_executive_insight(None, highest_site)

    return TrialMetrics(
        trialId=protocol.trialId,
        trialName=protocol.trialName,
        phase=protocol.phase,
        totalPatients=total_patients,
        totalSites=len(sites),
        totalDeviations=total_deviations,
        openDeviations=open_devs,
        criticalDeviations=critical_devs,
        majorDeviations=major_devs,
        trialRiskScore=trial_risk_score,
        trialRiskBand=trial_risk_band,
        compliancePercentage=compliance_pct,
        highRiskSitesCount=len(high_risk_sites),
        mediumRiskSitesCount=len(medium_risk_sites),
        siteRiskList=site_risk_list,
        executiveInsight=executive_insight,
    )
